cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,811 @@
+    col_ids.size(),
+    [node_categories = tree.node_categories.begin(),
+     col_ids         = col_ids.begin(),
+     row_offsets     = row_offsets.begin(),
+     range_begin     = tree.node_range_begin.begin(),
+     range_end       = tree.node_range_end.begin(),
+     d_ignore_vals   = d_ignore_vals.begin(),
+     d_columns_data  = d_columns_data.begin()] __device__(size_type i) {
+      switch (node_categories[i]) {
+        case NC_STRUCT: set_bit(d_columns_data[col_ids[i]].validity, row_offsets[i]); break;
+        case NC_LIST: set_bit(d_columns_data[col_ids[i]].validity, row_offsets[i]); break;
+        case NC_VAL:
+        case NC_STR:
+          if (d_ignore_vals[col_ids[i]]) break;
+          set_bit(d_columns_data[col_ids[i]].validity, row_offsets[i]);
+          d_columns_data[col_ids[i]].string_offsets[row_offsets[i]] = range_begin[i];
+          d_columns_data[col_ids[i]].string_lengths[row_offsets[i]] = range_end[i] - range_begin[i];
+          break;
+        default: break;
+      }
+    });
+
+  // 4. scatter List offset
+  //   sort_by_key {col_id}, {node_id}
+  //   unique_copy_by_key {parent_node_id} {row_offset} to
+  //   col[parent_col_id].child_offsets[row_offset[parent_node_id]]
+  rmm::device_uvector<NodeIndexT> original_col_ids(col_ids.size(), stream);  // make a copy
+  thrust::copy(rmm::exec_policy(stream), col_ids.begin(), col_ids.end(), original_col_ids.begin());
+  rmm::device_uvector<size_type> node_ids(row_offsets.size(), stream);
+  thrust::sequence(rmm::exec_policy(stream), node_ids.begin(), node_ids.end());
+  thrust::stable_sort_by_key(
+    rmm::exec_policy(stream), col_ids.begin(), col_ids.end(), node_ids.begin());
+  auto ordered_parent_node_ids =
+    thrust::make_permutation_iterator(tree.parent_node_ids.begin(), node_ids.begin());
+  auto ordered_row_offsets =
+    thrust::make_permutation_iterator(row_offsets.begin(), node_ids.begin());
+  thrust::for_each_n(
+    rmm::exec_policy(stream),
+    thrust::counting_iterator<size_type>(0),
+    col_ids.size(),
+    [num_nodes = col_ids.size(),
+     ordered_parent_node_ids,
+     ordered_row_offsets,
+     original_col_ids = original_col_ids.begin(),
+     col_ids          = col_ids.begin(),
+     row_offsets      = row_offsets.begin(),
+     node_categories  = tree.node_categories.begin(),
+     d_columns_data   = d_columns_data.begin()] __device__(size_type i) {
+      auto parent_node_id = ordered_parent_node_ids[i];
+      if (parent_node_id != parent_node_sentinel and node_categories[parent_node_id] == NC_LIST) {
+        // unique item
+        if (i == 0 ||
+            (col_ids[i - 1] != col_ids[i] or ordered_parent_node_ids[i - 1] != parent_node_id)) {
+          // scatter to list_offset
+          d_columns_data[original_col_ids[parent_node_id]]
+            .child_offsets[row_offsets[parent_node_id]] = ordered_row_offsets[i];
+        }
+        // TODO: verify if this code is right. check with more test cases.
```

Comment:
This should be correct if the ordered_parent_node_ids are also grouped together (not necessarily sorted, but all equal elements in the same range). I'd assume this is the case, but maybe you can confirm.